File: 1D Dynamic Programming/213. 打家劫舍 II/213. 打家劫舍 II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:return nums[0]
        return max(self.helper(nums[:-1]), self.helper(nums[1:]))

    def helper(self, nums):
        pre, cur = 0, 0

        #pre代表rob[i-2]、cur代表rob[i-1]
        for n in nums:
            next = max(pre + n, cur)
            pre = cur
            cur = next
        return cur

File: 1D Dynamic Programming/213. 打家劫舍 II/213. 打家劫舍 II.cpp
class Solution {

public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];

        vector<int> range1(nums.begin(), nums.end()-1);
        vector<int> range2(nums.begin() + 1, nums.end());

        return max(helper(range1), helper(range2));
    }

private:
    int helper(vector<int> nums){
        int pre = 0, cur = 0;
        int next;

        for(int i = 0; i < nums.size(); i++){
            next = max(nums[i] + pre, cur);
            pre = cur;
            cur = next;
        }
        return cur;
    }

};